File: Cpp/45_Jump_Game_II.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, INT_MAX);
        dp[0] = 0;
        for(int i=0; i<n; ++i){
            int max_jump = min(n, i + 1 + nums[i]); 
            for(int j=i+1; j<max_jump; ++j){
                dp[j] = min(dp[j], dp[i] + 1);
            }
        }
        
        return dp[n-1];
    }
};

File: Kotlin/45_Jump_Game_II.kt
class Solution {
    fun jump(nums: IntArray): Int {
        val n = nums.size
        val dp = IntArray(n) { Int.MAX_VALUE }
        dp[0] = 0

        for(i in 0 until n){
            val maxJump = minOf(n, i + 1 + nums[i])
            for(j in i + 1 until maxJump){
                dp[j] = minOf(dp[j], dp[i] + 1)
            }
        }
        
        return dp[n - 1]
    }
}